File: swift-prototype/Sources/BankingCore/TransferError.swift
// Protocolo de erro tipado — equivalente ao enum TransferError do Rust
public enum TransferError: Error, CustomStringConvertible {
    case insufficientFunds(available: Double, requested: Double)
    case accountNotFound(id: String)
    case negativeAmount(Double)
    case networkTimeout(retries: Int)
    case invalidLedgerState(String)

    public var description: String {
        switch self {
        case .insufficientFunds(let available, let requested):
            return String(format: "Saldo insuficiente: disponível R$%.2f, solicitado R$%.2f", available, requested)
        case .accountNotFound(let id):
            return "Conta não encontrada: \(id)"
        case .negativeAmount(let amount):
            return String(format: "Valor inválido: R$%.2f (deve ser positivo)", amount)
        case .networkTimeout(let retries):
            return "Timeout de rede após \(retries) tentativas"
        case .invalidLedgerState(let msg):
            return "Estado inválido do ledger: \(msg)"
        }
    }
}

File: java-prototype/src/TransferError.java
public abstract class TransferError extends Exception {
    public TransferError(String msg) {
        super(msg);
    }

    //saldo insuficiente
    public static class InsufficientFunds extends TransferError {
        public final double available;
        public final double requested;

        public InsufficientFunds( double available, double requested) {
            super("Saldo insuficiente: tem R$" + String.format("%.2f", available) + " mas tentou R$" + String.format("%.2f", requested));
            this.available = available;
            this.requested = requested;
        }
    }

    //conta inexistente no sistema
    public static class AccountNotFound extends TransferError {
        public final String accountId;

        public AccountNotFound(String id) {
            super("Conta nao encontrada: " + id);
            this.accountId = id;

        }
    }

    //valores invalidos - negativo, zero
    public static class NegativeAmount extends TransferError {
        public final double amount;
        public NegativeAmount(double amount) {
            super("Valor invalido: " + String.format("%.2f", amount));
            this.amount = amount;
        }
    }

    // timeout de rede apos N tentativas
    public static class NetworkTimeout extends TransferError {
        public final int retries;

        public NetworkTimeout(int retries){
            super("Timeout apos " + retries + " tentativas");
            this.retries = retries;
        }
    }

    // erros irrecuperaveis - em rust seria panic!
    public static class InvalidLedgerState extends RuntimeException {
        public InvalidLedgerState(String msg) {
            super("Estado invalido: " + msg);
        }
    }
}
